use-cases/client-server-msgs-perf/svmsg_file_server.c

```diff
@@ -456,8 +494,19 @@ svr_proc_process_msg(void *cfg)
     Client_info *clientp = NULL;
     requestMsg req;
 
+#if __APPLE__
+    uint64_t tid = svr_config->thread_idx;
+#else
+    pthread_t tid = pthread_self();
+#endif  // __APPLE__
+
+    uint64_t    num_ops = 0;
     for (;;) {
 
+        if (NumActiveClientsHWM && (NumActiveClients == 0)) {
+            goto end_forever_loop;
```

Comment:
Each thread will 'block' in `msgrcv()` on L510, below.

So, this check here is an early-check to see if all clients have drained out. If so, this thread can also exit.

Reply:
Is this a time-to-check-to-time-of-use race? Specifically, what happens if `NumActiveClients` becomes 0 after the check?

Reply:
The condition you described is handled by the "hacky-fix" I implemented around this block:

```
 565             if (NumActiveClients <= 0) {
 566                 printf("Server: ThreadID=%" PRIu64 " exiting.\n", tid);
 567
 568                 if (NumActiveClients == 0) {
 569                     req.mtype = REQ_MT_EXIT;
 570
 571                     // Inform last remaining server-thread that this is
 572                     // a final-exit from a server-thread, and not from a real
 573                     // client.
 574                     req.client_idx = -1;
 575                     if (msgsnd(serverId, &req, REQ_MSG_SIZE, 0) == -1) {
 576                         errExit("thread-msgsnd-exit");
 577                     }
 578                 }
```

The problem situation you describe should only occur when the last client exits the system; i.e. `NumActiveClients` transitions from 1 to 0.

- If thread-1 checks the variable and it's non-zero, the thread will end up inside `msgLen = msgrcv` on L515.
- When thread-2 detects on L568 above that the counter has gone to 0, it will send an `REQ_MT_EXIT` message on L575 (which is intended for the other thread-1 waiting for all clients to exit).
- Thread-1 will receive this message and will exit normally.

I have verified this thru test-execution and it seems to work -- at least, works sufficiently well-enough to get the client/server programs to exit cleanly so that perf-tests can be run.

Reply:
Ah, I get it now! Yes, makes sense.